App logic refactorization

albertinisg opened the pull request:
This PR fixes #76 and is based on the feedback received in #62 and #77. It includes the following changes:
- Fix/remove the 'promise hell', and remove the nested code
- Refactor the code used in several places
- Remove the 'filtering functions' and start using sql (simple query language) from Orion
- Decrease the number of requests against Orion
- New data image

Commit:
Update specs

Commit:
Update tourguide data

Commit:
Refactor application logic

jmcanterafonseca commented on server/feeders/reviewsgenerator.js:
```diff
@@ -139,22 +127,22 @@ var triggerRestaurantsRatings = function() {
-        reviews);
+  Object.keys(restaurantsData).forEach(function(element, index) {
+    var restaurantName = restaurantsData[index].name;
+    var sqlList = [];
```
can we name this `filter`, instead of sqlList? variable name is strange as it is not SQL actually

jmcanterafonseca commented on server/routes/orion.js:
```diff
@@ -32,231 +32,141 @@ tv4.addSchema('review', schema.review);
-    });
+  var restaurantId = utils.generateId(req.params.id);
+  var currentDate = new Date().toISOString();
+  var sqlList = [];
```
same comment

jmcanterafonseca commented on server/routes/orion.js:
```diff
@@ -32,231 +32,141 @@ tv4.addSchema('review', schema.review);
+  .then(function(reservations) {
+    occupancyLevelsObject = utils.updateOccupancyLevels(reservations.body,
+      currentDate);
+    return utils.getListByType('Restaurant', restaurantId, req.headers);
```
use a constant for the entity type 'Restaurant'. Same for the rest of entity types

jmcanterafonseca commented on server/routes/orion.js:
```diff
@@ -32,231 +32,141 @@ tv4.addSchema('review', schema.review);
+  var timeframe = utils.getTimeframe(req.params.date);
+  utils.addConditionToQuery(sqlList, 'startTime', '==', timeframe);
+  queryString.q = sqlList.join(';');
+  utils.getListByType('FoodEstablishmentReservation', null, req.headers,
```
entity type as constant

jmcanterafonseca commented on server/routes/orion.js:
```diff
@@ -265,541 +175,390 @@ exports.getOrganizationRestaurants = function(req, res) {
+                            req.params.restaurant);
+  queryString.q = sqlList.join(';');
+  utils.getListByType('FoodEstablishmentReservation', null, req.headers,
+                      queryString)
```
please be consistent with indentation, at some places you put the parameter which does not fit on the previous line just below the parenthesis above and in other places you do not do it. I would suggest all the code follows the same indentation convention, as here.

jmcanterafonseca commented on server/routes/orion.js:
```diff
@@ -265,541 +175,390 @@ exports.getOrganizationRestaurants = function(req, res) {
+    aggregateRatings = utils.getAggregateRating(reviews.body);
+    fiwareHeaders = utils.completeHeaders(req.headers, servicePath);
+    return utils.sendRequest('PATCH', aggregateRatings, restaurantId,
+      fiwareHeaders);
```
indentation, please see my comment above

jmcanterafonseca commented on server/routes/orion.js:
```diff
@@ -265,541 +175,390 @@ exports.getOrganizationRestaurants = function(req, res) {
+      aggregateRatings = utils.getAggregateRating(reviews.body);
+      fiwareHeaders = utils.completeHeaders(req.headers, servicePath);
+      return utils.sendRequest('PATCH', aggregateRatings, restaurantId,
+      fiwareHeaders);
```
indentation, see my comment below

jmcanterafonseca commented on server/routes/orion.js:
```diff
@@ -32,231 +32,141 @@ tv4.addSchema('review', schema.review);
+    servicePath = restaurant.body.department;
+    fiwareHeaders = utils.completeHeaders(req.headers, servicePath);
+    return utils.sendRequest('PATCH', occupancyLevelsObject, restaurantId,
+      fiwareHeaders);
```
indentation

jmcanterafonseca commented on server/routes/orion.js:
```diff
@@ -32,231 +32,141 @@ tv4.addSchema('review', schema.review);
+    servicePath = restaurant.body.department;
+    fiwareHeaders = utils.completeHeaders(req.headers, servicePath);
+    return utils.sendRequest('PATCH', occupancyLevelsObject, restaurantId,
+      fiwareHeaders);
```
indentation

jmcanterafonseca commented on server/utils.js:
```diff
@@ -706,59 +646,34 @@ function getAggregateRating(listOfReviews) {
-        listOfReservations[pos].reservationStatus == 'Confirmed') {
-      occupancyLevels += listOfReservations[pos].partySize;
-    }
+  Object.keys(listOfReservations).forEach(function(element, index) {
```
strange way to iterate over the list of reservations I think it should be 

``` js
listOfReservations.forEach(function(aReservation) {
}) 

```

jmcanterafonseca commented on server/feeders/reviewsgenerator.js:
```diff
@@ -139,22 +127,22 @@ var triggerRestaurantsRatings = function() {
-      restaurantReviews = utils.getRestaurantReviews(
-        restaurantName,
-        reviews);
+  Object.keys(restaurantsData).forEach(function(element, index) {
```
strange, strange, please check

jmcanterafonseca commented:
please check the latest comments, I think a new minor iteration is needed

Commit:
Minor bugfixes

This commit fix indentation, vars renaming and usage, as well as some
loops methods

albertinisg commented:
@jmcanterafonseca , I've addressed all your comments in the latest commit. Could you please also check PR's #79 and #80? Those changes are also needed for the correct functioning of Tourguide. Thanks!!

jmcanterafonseca commented on server/feeders/reviewsgenerator.js:
```diff
@@ -27,6 +27,8 @@ var reviewsAdded = 0;
 var delay = 200; //time in ms
 var restaurantsModified = 0;
+var rev = 'Review';
+var rstrnt = 'Restaurant';
```
use const and UPPERCASE

const RESTAURANT_TYPE = 'Restaurant'
const REVIEW_TYPE = 'Review'

albertinisg replied:
@jmcanterafonseca I'm having issues with this and the jshint validators. By default, we set `/* jshint node:true */` at the very beggining of each file. `const` is available in ES6, but adding `jshint esversion: 6` does not solve the issue. After reading a bit, it seems that in node, a `const` shouldn't be declared using `const` but `var` (in uppercase) (see [node-style-guide](https://github.com/felixge/node-style-guide#use-uppercase-for-constants)). So I suggest to declare it as follows:

``` node
var RESTAURANT_TYPE = 'Restaurant;'
var REVIEW_TYPE = 'Review';
```

jmcanterafonseca replied:
ok

jmcanterafonseca commented on server/feeders/reviewsgenerator.js:
```diff
@@ -139,22 +129,21 @@ var triggerRestaurantsRatings = function() {
-      q.push({'aggregateRatings': ratings,
-        'restaurantName': restaurantName}, returnPost);
+  restaurantsData.forEach(function(element, index) {
+    var restaurantName = restaurantsData[index].name;
```
you don't really need the index here, forEach gives you each array element

jmcanterafonseca commented on server/feeders/reviewsgenerator.js:
```diff
@@ -59,16 +61,16 @@ var feedOrionReviews = function() {
   };
 
-  Object.keys(restaurantsData).forEach(function(element, index) {
+  restaurantsData.forEach(function(element, index) {
```
don't need the index

jmcanterafonseca commented on server/utils.js:
```diff
@@ -454,9 +454,9 @@ function dataToSchema(listOfElements) {
   listOfElements = objectToArray(listOfElements);
 
-  Object.keys(listOfElements).forEach(function(element, pos) {
+  listOfElements.forEach(function(element, index) {
```
don't need the index

jmcanterafonseca commented on server/utils.js:
```diff
@@ -691,7 +631,7 @@ function getAggregateRating(listOfReviews) {
   listOfReviews = objectToArray(listOfReviews);
 
-  Object.keys(listOfReviews).forEach(function(element, index) {
+  listOfReviews.forEach(function(element, index) {
```
don't need the index

jmcanterafonseca commented:
some tiny fixes still required I think this is the last round

Commit:
Remove index in forEach loops and use uppercase for consts

albertinisg commented:
@jmcanterafonseca tiny fixes done

jmcanterafonseca commented:
LGTM, thanks for the work!
